staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy.go

```diff
@@ -17,51 +17,128 @@ limitations under the License.
 	serverId string,
+	identityLeaseLabelSelector string,
+	leaseInformer coordinationv1informers.LeaseInformer,
 	reconciler reconcilers.PeerEndpointLeaseReconciler,
-	serializer runtime.NegotiatedSerializer) *peerProxyHandler {
+	ser runtime.NegotiatedSerializer,
+	loopbackClientConfig *rest.Config,
+	proxyClientConfig *transport.Config,
+) (*peerProxyHandler, error) {
+	if parts := strings.Split(identityLeaseLabelSelector, "="); len(parts) != 2 {
+		return nil, fmt.Errorf("invalid identityLeaseLabelSelector provided, must be of the form key=value, received: %v", identityLeaseLabelSelector)
+	}
+
+	selector, err := labels.Parse(identityLeaseLabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse label selector: %w", err)
+	}
+
 	h := &peerProxyHandler{
-		name:                  "PeerProxyHandler",
-		storageversionManager: svm,
-		proxyTransport:        proxyTransport,
-		svMap:                 sync.Map{},
-		serverId:              serverId,
-		reconciler:            reconciler,
-		serializer:            serializer,
+		name:                      "PeerProxyHandler",
+		serverId:                  serverId,
+		reconciler:                reconciler,
+		serializer:                ser,
+		loopbackClientConfig:      loopbackClientConfig,
+		proxyClientConfig:         proxyClientConfig,
+		localDiscoveryInfoCache:   make(map[schema.GroupVersionResource]bool),
+		localDiscoveryCacheTicker: time.NewTicker(localDiscoveryRefreshInterval),
+		peerDiscoveryInfoCache:    make(map[string]*peerDiscoveryInfo),
+		peerLeaseQueue:            workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[string](), workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName}),
+		apiserverIdentityInformer: leaseInformer,
 	}
-	svi := informerFactory.Internal().V1alpha1().StorageVersions()
-	h.storageversionInformer = svi.Informer()
 
-	svi.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+	h.identityLeaseLabelSelector = selector
+	discoveryScheme := runtime.NewScheme()
+	utilruntime.Must(apidiscoveryv2.AddToScheme(discoveryScheme))
+	h.discoverySerializer = serializer.NewCodecFactory(discoveryScheme)
+
+	peerDiscoveryRegistration, err := h.apiserverIdentityInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			h.addSV(obj)
+			if lease, ok := h.isValidPeerIdentityLease(obj); ok {
+				h.enqueueLease(lease)
+			}
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			h.updateSV(oldObj, newObj)
+			oldLease, oldLeaseOk := h.isValidPeerIdentityLease(oldObj)
+			newLease, newLeaseOk := h.isValidPeerIdentityLease(newObj)
+			if oldLeaseOk && newLeaseOk &&
+				oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+				h.enqueueLease(newLease)
+			}
 		},
 		DeleteFunc: func(obj interface{}) {
-			h.deleteSV(obj)
-		}})
-	return h
+			if lease, ok := h.isValidPeerIdentityLease(obj); ok {
```

Comment:
Delete is a bit more convoluted to get the object, maybe you can just use this inside `isValidPeerIdentityLease` a
```
	lease, ok := obj.(*v1.Lease)
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			utilruntime.HandleError(fmt.Errorf("unexpected object type: %T", obj))
			return
		}
		if lease, ok = tombstone.Obj.(*v1.Lease); !ok {
			utilruntime.HandleError(fmt.Errorf("unexpected object type: %T", obj))
			return
		}
	}
```

Reply:
Done